frontend/model/contracts/voting/proposals.js

```diff
@@ -0,0 +1,171 @@
+'use strict'
+
+import sbp from '~/shared/sbp.js'
+// import Vue from 'vue'
+import { objectOf, literalOf, unionOf, number } from '~/frontend/utils/flowTyper.js'
+import { DAYS_MILLIS } from '~/frontend/utils/time.js'
+import { PROPOSAL_RESULT } from '~/frontend/utils/events.js'
+import rules, { ruleType, VOTE_UNDECIDED, VOTE_AGAINST, VOTE_FOR, RULE_THRESHOLD, RULE_DISAGREEMENT } from './rules.js'
+
+export const PROPOSAL_INVITE_MEMBER = 'invite-member'
+export const PROPOSAL_REMOVE_MEMBER = 'remove-member'
+export const PROPOSAL_GROUP_SETTING_CHANGE = 'group-setting-change'
+export const PROPOSAL_PROPOSAL_SETTING_CHANGE = 'proposal-setting-change'
+export const PROPOSAL_GENERIC = 'generic'
+
+export const STATUS_OPEN = 'open'
+export const STATUS_PASSED = 'passed'
+export const STATUS_FAILED = 'failed'
+export const STATUS_EXPIRED = 'expired'
+export const STATUS_WITHDRAWN = 'withdrawn'
+
+export function archiveProposal (state, proposalHash) {
+  // TODO: handle this better (archive the proposal or whatever)
+  console.warn(`archiveProposal is not fully implemented yet...`)
+  // Vue.delete(state.proposals, proposalHash)
+}
+
+export const proposalSettingsType = objectOf({
+  rule: ruleType,
+  expires_ms: number,
+  ruleSettings: objectOf({
+    [RULE_THRESHOLD]: objectOf({ threshold: number }),
+    [RULE_DISAGREEMENT]: objectOf({ threshold: number })
+  })
+})
+
+// returns true IFF a single YES vote is required to pass the proposal
+export function oneVoteToPass (proposalHash) {
+  const rootState = sbp('state/vuex/state')
+  const state = rootState[rootState.currentGroupId]
+  const proposal = state.proposals[proposalHash]
+  const votes = Object.assign({}, proposal.votes)
+  const currentResult = rules[proposal.data.votingRule](state, proposal.data.proposalType, votes)
+  votes[String(Math.random())] = VOTE_FOR
+  const newResult = rules[proposal.data.votingRule](state, proposal.data.proposalType, votes)
+  console.debug(`oneVoteToPass currentResult(${currentResult}) newResult(${newResult})`)
+  return currentResult === VOTE_UNDECIDED && newResult === VOTE_FOR
+}
+
+function voteAgainst (state, data) {
+  const { proposalHash } = data
+  const proposal = state.proposals[proposalHash]
+  proposal.status = STATUS_FAILED
+  archiveProposal(state, proposalHash)
+  sbp('okTurtles.events/emit', PROPOSAL_RESULT, state, VOTE_AGAINST, data)
+}
+
+const proposals = {
+  [PROPOSAL_INVITE_MEMBER]: {
+    defaults: {
+      rule: RULE_THRESHOLD, // the default voting rule governing invites
+      expires_ms: 14 * DAYS_MILLIS,
+      ruleSettings: {
+        [RULE_THRESHOLD]: { threshold: 0.8 },
+        [RULE_DISAGREEMENT]: { threshold: 1 }
+      }
+    },
+    [VOTE_FOR]: function (state, data) {
+      const proposal = state.proposals[data.proposalHash]
+      proposal.payload = data.passPayload
+      proposal.status = STATUS_PASSED
+      sbp('gi.contracts/group/invite/process', state, {
+        meta: proposal.meta,
```

Comment:
Why passing `meta`? It seems is not used when received at `group.js`?

Reply:
It is used by various UI widgets, for example have a look at `ProposalBox.vue`.

Reply:
Erm, I was mistaken and you are right. It's not currently used by `'gi.contracts/group/invite'`, however all `group` messages are supposed to have metadata associated with them, so I pass along the metadata just in case it ever is used.

Reply:
I see... but if it's not needed now, why passing "noise". We should keep it clear/simple/pass only the essential. Anyway, could you explain to me a little more about what this `metadata` could contain?

Reply:
It's not noise at all, it's part of what is essential. Since the group contract defines metadata, it means every message must have metadata attached to it. If I were to delete this line, it would throw an exception when the message is validated. Look at the group contract definition in `group.js` to see what `metadata` must contain.

Reply:
It needs `metadata` to create a group, but it doesn't need to invite someone... But I understand your point.

Resolving this.